### DISCUSSION TITLE
How to set permission to INSERT from trigger

### DISCUSSION THREAD

**Discussion by Falci** [supabase/supabase#18157] [Questions] [answered]
Hey!

Following my [previous question](https://github.com/orgs/supabase/discussions/18139), I have this scenario:

1. A table `public.users`:
id, name, avatar, createdAt, updatedAt

2. a trigger on `BEFORE INSERT` for `auth.users` where I validate if the user email is present in a 3rd-party app (via `http_get`).

I'd like to update this trigger to also populate the `public.users`, but I'm getting "permission denied"

```sql
CREATE OR REPLACE FUNCTION public.copy_users()
RETURNS trigger
LANGUAGE plpgsql
AS $$
BEGIN

  INSERT INTO public.users (id, name, avatar, "createdAt", "updatedAt")
  VALUES ((NEW.raw_user_meta_data->>'user_id')::int, NEW.raw_user_meta_data->>'full_name', NEW.raw_user_meta_data->>'picture', now(), now())
  ON CONFLICT (id) DO UPDATE SET
    name = NEW.raw_user_meta_data->>'full_name',
    avatar = NEW.raw_user_meta_data->>'picture',
    "updatedAt" = now();

  RETURN NEW;
end;
$$;

CREATE TRIGGER copy_users_trigger
AFTER INSERT ON auth.users
FOR EACH ROW EXECUTE FUNCTION public.copy_users();
```

How can I create a permission that allows the trigger to write in the `public.users` table?

**Comment by GaryAustin1** [comment, marked as answer]
You just need to add `security definer` after `language plpgsql`.
If you look in the SQL editor at Quickstarts and user management starter you can see a template for the function.
The auth user does not have access to public, so security definer runs as postgres user which does.

**Reply by Falci** [reply]
```sql
CREATE OR REPLACE FUNCTION public.copy_users()
RETURNS trigger
AS $$
BEGIN

  INSERT INTO public.users (id, name, avatar, "createdAt", "updatedAt")
  VALUES ((NEW.raw_user_meta_data->>'user_id')::int, NEW.raw_user_meta_data->>'full_name', NEW.raw_user_meta_data->>'picture', now(), now())
  ON CONFLICT (id) DO UPDATE SET
    name = NEW.raw_user_meta_data->>'full_name',
    avatar = NEW.raw_user_meta_data->>'picture',
    "updatedAt" = now();

  RETURN NEW;
end;
$$ LANGUAGE plpgsql security definer;
```

**Reply by Falci** [reply]
For future reference: [link](https://supabase.com/docs/guides/auth/managing-user-data#using-triggers)

**Reply by devhandler** [reply]
@GaryAustin1 security definer will definitely work, but with rls enabled and insert policy check expression: (auth.uid() = id), why does security invoker not work? the user who inserts their new record to public.profile should be able to pass the rls insert check but i still get denied access to profile.   their auth.uid() match their id value in the new row, right? confused...

**Reply by GaryAustin1** [reply]
@devhandler  The example the user presented is auth.users insert from gotrue.   Gotrue user role does not have access to the public schema.  That role is not the API user role as the user may not even be confirmed yet.

**Reply by devhandler** [reply]
Makes sense. But my case is a bit different. I'm following the example on supabase doc. Inserting a new user record that is triggered AFTER auth.users insert, in which case the new user should already be in auth.users table. But I still get denied permission with profile insert policy check auth.uid()=id.  

Ofc using security definer solves it, I just don't understand why security invoker doesn't work in this case.

**Reply by GaryAustin1** [reply]
Is it triggered on the signup operation and initially started with the auth.users insert?  If so the auth user can't insert into the table.  It is likely not an RLS thing, but a permission to public schema thing.  It would be better if you started a new post here or Discord and show what you are up to.

**Reply by devhandler** [reply]
sg. thank you so much for the help! i have been learning a lot from you.
